lib/actions/device.js

```diff
@@ -0,0 +1,535 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import * as commandOptions from './command-options';
+
+import * as _ from 'lodash';
+import { normalizeUuidProp } from '../utils/normalization';
+import { getBalenaSdk, getVisuals } from '../utils/lazy';
+
+/** @type {import('balena-sdk').PineOptionsFor<import('balena-sdk').Device>} */
+const expandForAppName = {
+	$expand: { belongs_to__application: { $select: 'app_name' } },
+};
+
+export const list = {
+	signature: 'devices',
+	description: 'list all devices',
+	help: `\
+Use this command to list all devices that belong to you.
+
+You can filter the devices by application by using the \`--application\` option.
+
+Examples:
+
+	$ balena devices
+	$ balena devices --application MyApp
+	$ balena devices --app MyApp
+	$ balena devices -a MyApp\
+`,
+	options: [commandOptions.optionalApplication],
+	permission: 'user',
+	primary: true,
+	action(_params, options) {
+		const Promise = require('bluebird');
+		const balena = getBalenaSdk();
+
+		return Promise.try(function() {
+			if (options.application != null) {
+				return balena.models.device.getAllByApplication(
+					options.application,
+					expandForAppName,
+				);
+			}
+			return balena.models.device.getAll(expandForAppName);
+		}).tap(function(devices) {
+			devices = _.map(devices, function(device) {
+				// @ts-ignore extending the device object with extra props
+				device.dashboard_url = balena.models.device.getDashboardUrl(
+					device.uuid,
+				);
+				// @ts-ignore extending the device object with extra props
+				device.application_name = device.belongs_to__application?.[0]
+					? device.belongs_to__application[0].app_name
+					: 'N/a';
+				device.uuid = device.uuid.slice(0, 7);
+				return device;
+			});
+
+			console.log(
+				getVisuals().table.horizontal(devices, [
+					'id',
+					'uuid',
+					'device_name',
+					'device_type',
+					'application_name',
+					'status',
+					'is_online',
+					'supervisor_version',
+					'os_version',
+					'dashboard_url',
+				]),
+			);
+		});
+	},
+};
+
+export const info = {
+	signature: 'device <uuid>',
+	description: 'list a single device',
+	help: `\
+Use this command to show information about a single device.
+
+Examples:
+
+	$ balena device 7cf02a6\
+`,
+	permission: 'user',
+	primary: true,
+	action(params) {
+		normalizeUuidProp(params);
+		const balena = getBalenaSdk();
+
+		return balena.models.device
+			.get(params.uuid, expandForAppName)
+			.then(device =>
+				// @ts-ignore `device.getStatus` requires a device with service info, but
+				// this device isn't typed with them, possibly needs fixing?
+				balena.models.device.getStatus(device).then(function(status) {
```

Comment:
@thgreasi any idea whether this is actually valid or not? Typescript complains about it but I assume it must work or there'd be bug reports against the coffee version?